cmd/data-usage.go

```diff
@@ -177,10 +181,92 @@ func loadDataUsageFromBackend(ctx context.Context, objAPI ObjectLayer) (DataUsag
 type Item struct {
 	Path string
 	Typ  os.FileMode
+
+	bucket     string // Bucket.
+	prefix     string // Only the prefix if any, does not have final object name.
+	objectName string // Only the object name without prefixes.
+	lifeCycle  *lifecycle.Lifecycle
+	debug      bool
 }
 
 type getSizeFn func(item Item) (int64, error)
 
+// actionMeta contains information used to apply actions.
+type actionMeta struct {
+	oi      ObjectInfo
+	trustOI bool // Set true if oi can be trusted and has been read with quorum.
+	meta    map[string]string
+}
+
+// applyActions will apply lifecycle checks on to a scanned item.
+// The resulting size on disk will always be returned.
+// The metadata will be compared to consensus on the object layer before any changes are applied.
+// If no metadata is supplied, -1 is returned if no action is taken.
+func (i *Item) applyActions(ctx context.Context, o ObjectLayer, meta actionMeta) (size int64) {
+	size, err := meta.oi.GetActualSize()
+	if i.debug {
+		logger.LogIf(ctx, err)
+	}
+	if i.lifeCycle == nil {
+		// logger.Info(color.Green("applyActions:")+" %q: no lifecycle", i.objectPath())
+		return size
+	}
+
+	action := i.lifeCycle.ComputeAction(i.objectPath(), meta.meta[xhttp.AmzObjectTagging], meta.oi.ModTime)
+	logger.Info(color.Green("applyActions:")+" lifecycle: %q, initial scan: %v", i.objectPath(), action)
+	switch action {
+	case lifecycle.DeleteAction:
+	default:
+		// No action.
+		return size
+	}
+
+	// These (expensive) operations should only run on items we are likely to delete.
+	// Load to ensure that we have the correct version and not an unsynced version.
+	if !meta.trustOI {
+		obj, err := o.GetObjectInfo(ctx, i.bucket, i.objectPath(), ObjectOptions{})
+		if err != nil {
+			// Do nothing - heal in the future.
+			logger.LogIf(ctx, err)
+			return size
+		}
+		size = obj.Size
+
+		// Recalculate action.
+		action = i.lifeCycle.ComputeAction(i.objectPath(), obj.UserTags, obj.ModTime)
+		logger.Info(color.Green("applyActions:")+" lifecycle: secondary scan: %v", action)
```

Comment:
we would have to put this in `if i.debug`

Reply:
Same.